iroha_data_model/src/events.rs

```diff
@@ -228,8 +230,122 @@ pub mod pipeline {
         Transaction,
     }
 
+    /// Transaction was reject during verification of signature
+    #[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize, Decode, Encode)]
+    pub struct SignatureVerificationFail {
+        /// Signature which verification has failed
+        pub signature: Signature,
+        /// Error which happened during verification
+        pub reason: String,
+    }
+
+    /// Transaction was reject because it doesn't satisfy signature condition
+    #[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize, Decode, Encode)]
+    pub struct UnsatisfiedSignatureConditionFail {
+        /// Reason why signature condition failed
+        pub reason: String,
+    }
+
+    /// Transaction was reject because of fail of instruction
+    #[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize, Decode, Encode)]
+    pub struct InstructionExecutionFail {
+        /// Instruction which execution failed
+        pub instruction: Instruction,
+        /// Error which happened during execution
+        pub reason: String,
+    }
+
+    /// Transaction was reject because of low authority
+    #[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize, Decode, Encode)]
+    pub struct NotPermittedFail {
+        /// Reason of failure
+        pub reason: String,
+    }
+
+    /// The reason for rejecting transaction which happened because of new blocks.
+    #[derive(Debug, Clone, Copy, Eq, PartialEq, Serialize, Deserialize, Decode, Encode)]
+    pub enum BlockRejectionReason {
+        /// Block was rejected during consensus.
+        //TODO: store rejection reasons for blocks?
+        ConsensusBlockRejection,
+    }
+
+    /// The reason for rejecting transaction which happened because of transaction.
+    #[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize, Decode, Encode)]
+    pub enum TransactionRejectionReason {
+        /// Failed due to low authority.
+        NotPermitted(NotPermittedFail),
+        /// Failed verify signature condition specified in the account.
+        UnsatisfiedSignatureCondition(UnsatisfiedSignatureConditionFail),
+        /// Failed execute instruction.
+        InstructionExecution(InstructionExecutionFail),
+        /// Failed to verify signatures.
+        SignatureVerification(SignatureVerificationFail),
+        /// Genesis account can sign only transactions in the genesis block.
+        UnexpectedGenesisAccountSignature,
+    }
+
     /// The reason for rejecting transaction.
-    pub type RejectionReason = String;
+    #[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize, Decode, Encode)]
+    pub enum RejectionReason {
```

Comment:
Also `FromVariant` here. Would be actually cool to have `FromVariant` for N levels of enums, but that's just a thought.